sdb/commands/zfs/internal/__init__.py

```diff
@@ -81,6 +85,188 @@ def WEIGHT_GET_COUNT(weight: int) -> int:
     return BF64_GET((weight), 0, 54)
 
 
+def BPE_GET_ETYPE(bp: drgn.Object) -> int:
+    return BF64_GET(bp.blk_prop, 40, 8)
+
+
+def BPE_GET_LSIZE(bp: drgn.Object) -> int:
+    return BF64_GET_SB(bp.blk_prop, 0, 25, 0, 1)
+
+
+def BPE_GET_PSIZE(bp: drgn.Object) -> int:
+    return BF64_GET_SB(bp.blk_prop, 25, 7, 0, 1)
+
+
+def BP_GET_LSIZE(bp: drgn.Object) -> int:
+    if BP_IS_EMBEDDED(bp):
+        if BPE_GET_ETYPE(bp) == BP_EMBEDDED_TYPE_DATA:
+            return BPE_GET_LSIZE(bp)
+        return 0
+    return BF64_GET_SB(bp.blk_prop, 0, SPA_LSIZEBITS, SPA_MINBLOCKSHIFT, 1)
+
+
+def BP_GET_PSIZE(bp: drgn.Object) -> int:
+    if BP_IS_EMBEDDED(bp):
+        return 0
+    return BF64_GET_SB(bp.blk_prop, 16, SPA_PSIZEBITS, SPA_MINBLOCKSHIFT, 1)
+
+
+def BP_GET_COMPRESS(bp: drgn.Object) -> int:
+    return BF64_GET(bp.blk_prop, 32, SPA_COMPRESSBITS)
+
+
+def BP_IS_EMBEDDED(bp: drgn.Object) -> bool:
+    return bool(BF64_GET(bp.blk_prop, 39, 1))
+
+
+def BP_GET_CHECKSUM(bp: drgn.Object) -> int:
+    if BP_IS_EMBEDDED(bp):
+        return ZIO_CHECKSUM_OFF
+    return BF64_GET(bp.blk_prop, 40, 8)
+
+
+def BP_GET_TYPE(bp: drgn.Object) -> int:
+    return BF64_GET(bp.blk_prop, 48, 8)
+
+
+def BP_GET_LEVEL(bp: drgn.Object) -> int:
+    return BF64_GET(bp.blk_prop, 56, 5)
+
+
+def BP_USES_CRYPT(bp: drgn.Object) -> bool:
+    return bool(BF64_GET(bp.blk_prop, 61, 1))
+
+
+def BP_IS_ENCRYPTED(bp: drgn.Object) -> bool:
+    return (BP_USES_CRYPT(bp) and BP_GET_LEVEL(bp) <= 0 and
+            DMU_OT_IS_ENCRYPTED(BP_GET_TYPE(bp)))
+
+
+def BP_IS_AUTHENTICATED(bp: drgn.Object) -> bool:
+    return (BP_USES_CRYPT(bp) and BP_GET_LEVEL(bp) <= 0 and
+            not DMU_OT_IS_ENCRYPTED(BP_GET_TYPE(bp)))
+
+
+def BP_HAS_INDIRECT_MAC_CKSUM(bp: drgn.Object) -> bool:
+    return (BP_USES_CRYPT(bp) and BP_GET_LEVEL(bp) > 0)
+
+
+def BP_GET_DEDUP(bp: drgn.Object) -> bool:
+    return bool(BF64_GET(bp.blk_prop, 62, 1))
+
+
+def BP_GET_BYTEORDER(bp: drgn.Object) -> int:
+    return BF64_GET(bp.blk_prop, 63, 1)
+
+
+def BP_GET_LAYER(bp: drgn.Object) -> int:
+    if BP_IS_EMBEDDED(bp):
+        if sdb.get_type('blkptr_t').has_member('blk_logical_birth'):
+            return BF64_GET(bp.blk_logical_birth, 56, 8)
+        return BF64_GET(bp.blk_birth, 56, 8)
+    if sdb.get_type('blkptr_t').has_member('blk_physical_birth'):
+        return BF64_GET(bp.blk_physical_birth, 56, 8)
```

Comment:
the layer info might be in the logical birth word, or it might be in the physical birth word?

Reply:
The layer is in both birth words if they exist (i.e. embedded blocks don't have physical birth). This is really just for consistency but I can change this macro to always look at the logical birth word since that should always exist.